dma-trace: Full support for re-configuration and make dmatb persistent from the time it is allocated

ujfalusi opened the pull request:
Hi,

This PR is replacing #4879.

Since last time I have noticed another path for a memory leak which was a hard one to crack:
Within dma_trace_start() when CONFIG_DMA_GW is set we were allocating a dma_sg (to SOF_MEM_ZONE_SYS) and we would do that for every start. The config is only used in that function to configure the DMA channel, but the dma_sg can not be freed up as it is allocated to SOF_MEM_ZONE_SYS. Other zone allocation failed on my tgl-h laptop.

The solution is to make dmatb and this dma_sg persistent from the point they are allocated.
This comes with a bonus that we can still collect dtrace logs to SOF firmware for some extent (if the would be dtrace client is removed we would still have some memory of the events).

The other change compared to #4879 is that I don't reset the firmware dmatb, only the host pointers as when the re-configuration happens it is indicating that  we are going to have a new host destination along with 0 offsets.

lyakh commented on src/trace/dma-trace.c:
```diff
@@ -291,6 +302,8 @@ static int dma_trace_start(struct dma_trace_data *d)
 	}
 
 	if (d->dc.chan) {
+		unsigned int flags;
```
just noticed, that this file uses `unsigned int` for spin-lock flags. It should actually be `uint32_t`. Not something to change in this PR, but a separate PR could update that

ujfalusi replied:
I'll fix it up.

ujfalusi replied:
Replaced with `k_spinlock_key_t` as we have switched lock API...

lyakh commented on src/trace/dma-trace.c:
```diff
@@ -311,6 +324,14 @@ static int dma_trace_start(struct dma_trace_data *d)
+		spin_lock_irq(&d->lock, flags);
+
+		d->old_host_offset = 0;
+		d->posn.host_offset = 0;
```
above `dma_stop()` or `dma_copy_set_stream_tag()` could've failed. Don't you need to abort in some of those cases? I'd reformat this function with the usual `goto`s to make error handling more obvious

ujfalusi replied:
@lyakh, we do abort at line 339.

This is a bit tricky. I would say that we can (and need) to reset the the host_offset in any case, so this can be moved beyond line 341.

ujfalusi replied:
Moved after the error check/abort.

marc-hb commented:
SOFCI TEST

jsarha commented on src/trace/dma-trace.c:
```diff
@@ -228,7 +228,7 @@ static int dma_trace_buffer_init(struct dma_trace_data *d)
 		      DMA_TRACE_LOCAL_SIZE);
 	if (!buf) {
-		tr_err(&dt_tr, "dma_trace_buffer_init(): alloc failed");
+		mtrace_printf(LOG_LEVEL_ERROR, "dma_trace_buffer_init(): alloc failed");
```
Could start to use "%s(): ...", __func__ now that you touch the line.

jsarha reviewed (approved):
Only only one minor and optional nitpick. Otherwise looks good to me, but you'll need to rebase this after my dtrace buffer alignment PR.

ujfalusi commented:
Changes since v1 (I think):
- rebased on main
- addressed comments from @lyakh

lgirdwood reviewed (commented):
@ujfalusi some red showing on Zephyr APL, can you split this PR into more parts, based on risk and change size. This way it should be more obvious which parts are causing the CI to report issues. Thanks.

ujfalusi commented:
> @ujfalusi some red showing on Zephyr APL, can you split this PR into more parts, based on risk and change size. This way it should be more obvious which parts are causing the CI to report issues. Thanks.

@lgirdwood, we have exactly the same failures present in unrelated kernel pr, like this: https://sof-ci.01.org/linuxpr/PR3424/build7156/devicetest/

I can split the PR into two, but they can not go in parallel as the patches are depending on each other.

lgirdwood commented:
@ujfalusi no need to split, just seen the other PR showing red too in the same place for APL Zephyr.

lyakh commented on src/trace/dma-trace.c:
```diff
@@ -344,6 +356,14 @@ static int dma_trace_start(struct dma_trace_data *d)
+	d->old_host_offset = 0;
+	d->posn.host_offset = 0;
+
+	k_spin_unlock(&d->lock, key);
```
sorry, not sure I understand the use of locking well here. I looked through dma-trace.c and I couldn't understand what this lock is actually protecting. As for these `host_offset` fields - looks like in other locations they are accessed and modified without holding the lock?

ujfalusi replied:
True, there is no need for the lock.

ujfalusi replied:
yes, I will.

lyakh commented on src/trace/dma-trace.c:
```diff
@@ -218,8 +222,26 @@ int dma_trace_host_buffer(struct dma_trace_data *d,
+	key = k_spin_lock(&d->lock);
+
+	rfree(buffer->addr);
+	memset(buffer, 0, sizeof(*buffer));
```
why is this needed? I see, that it is just moved from another location, just wondering if we know why we need this?

ujfalusi replied:
It is to reset the variables within buffer.

lgirdwood commented:
SOFCI TEST

lgirdwood commented:
Rerun the CI since this is a trace change.

marc-hb commented on src/trace/dma-trace.c:
```diff
@@ -260,28 +293,63 @@ static int dma_trace_buffer_init(struct dma_trace_data *d)
+	 */
+#define SOF_BANNER_COMMON  \
+	"FW ABI 0x%x DBG ABI 0x%x tags SOF:" SOF_GIT_TAG  ZEPHYR_VER_OPT  \
+	" src hash 0x%08x (ldc hash " META_QUOTE(SOF_SRC_HASH) ")"
```
Can you please move these macros down, back together with the only place they are used?

marc-hb reviewed (changes requested):
I wonder what effect it will have on
- #4334 
I haven't found the time for a full review yet sorry.

ujfalusi commented:
> I wonder what effect it will have on
> 
>     * #4334
> 
> 
> I haven't found the time for a full review yet sorry.

I think a simple revert as it is in #4334 will no longer going to cut it to have early dtrace buffer as the `dma_trace_buffer_init()` now needs d->dc.dmac which is initialized in `dma_trace_init_complete()`.
#4334 also does not address the memory leak this PR is fixing.

We should revisit the early dtrace buffer allocation, I agree. I would also add a Kconfig option to make it selectable with opt-in. I would also add mtrace for wrap notification when the host is not yet (or withdrawn it) provided a host buffer to flush the dtrace buffer to.

marc-hb commented:
SOFCI TEST

ujfalusi commented:
Changes since v2:
- drop the pointless locking as pointed out by @lyakh 
- DO not separate the macro used from the tag print from the printing, suggested by @marc-hb

ujfalusi commented:
I have found other memory leak. I will update this PR with an additional patch (the host buffer's elem_array is not freed on error and on dtrace_free)

ujfalusi commented:
Changes since v3:
- two new patches added to fix other memory leaks:
  - `ipc: ipc3: handler: Fix error handling in ipc_dma_trace_config()`
  - `dma-trace: Fix memory leak of host elem_array on reconfiguration`
